File: statemachine/src/main/java/org/nichel/sm/Action.java
package org.nichel.sm;

import org.nichel.sm.annotation.OnPerformAction;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class Action {
    private final Object target;
    private final Method[] methods;

    final Status startStatus;
    final Status finalStatus;

    public Action(final Object target, final Status s1, final Status s2) {
        this.target = target;
        this.methods = target.getClass().getDeclaredMethods();
        this.startStatus = s1;
        this.finalStatus = s2;
    }

    Status perform(final StateMachine stateMachine) throws IllegalAccessException, InvocationTargetException, InstantiationException {
        onPerformAction(stateMachine, this);

        return finalStatus;
    }

    private Status onPerformAction(final StateMachine stateMachine, final Action action) throws IllegalAccessException, InstantiationException, InvocationTargetException {
        for (final Method method : methods) {
            if (method.isAnnotationPresent(OnPerformAction.class)) {
                final OnPerformAction annotation = method.getAnnotation(OnPerformAction.class);

                if (annotation.enter().equals(action.startStatus.getLabel()) && annotation.exit().equals(action.finalStatus.getLabel())) {
                    method.invoke(target, stateMachine, action);
                }
            }
        }

        return action.finalStatus;
    }

    @Override
    public String toString() {
        return startStatus.toString() + " -> " + finalStatus.toString();
    }
}

File: statemachine/src/main/java/org/nichel/sm/Action.kt
package org.nichel.sm

import org.nichel.sm.annotation.OnPerformAction
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*

public class Action(_target: Any, _enterStatus: Status, _exitStatus: Status) {
    private val target = _target
    internal val enterStatus = _enterStatus
    internal val exitStatus =_exitStatus

    private val onPerformMethods = ArrayList<Method>()

    init {
        for (method in target.javaClass.declaredMethods) {
            if (method.isAnnotationPresent(OnPerformAction::class.java)) {
                val annotation = method.getAnnotation(OnPerformAction::class.java)

                if (annotation.enter == enterStatus.label && annotation.exit == exitStatus.label) {
                    onPerformMethods.add(method)
                }
            }
        }
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    private fun onPerformAction(stateMachine: StateMachine, action: Action): Status {
        for (method in onPerformMethods) {
            method.invoke(target, stateMachine, action)
        }

        return action.exitStatus
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    internal fun perform(stateMachine: StateMachine): Status {
        return onPerformAction(stateMachine, this)
    }
}
